### PR TITLE
Add support for the new VQE workflows to the differentiable batch execution pipeline

### PR DESCRIPTION
**Pull request by josh146**
**Context:** In #1483, #1596, and #1551, support was added for:

- Directly specifying `expval(H)` inside of a QNode
- Differentiating Hamiltonian coefficients
- Allowing certain devices, including `default.qubit`, to support Hamiltonians directly.

However, these changes were not ported over to the new differentiable batch-execute pipeline.

**Description of the Change:**

- Add support for differentiating (one or more) Hamiltonians to `qml.gradients.param_shift`.

- Modify `hamiltonian_grad` to support multiple Hamiltonians (`return expval(H1), expval(H2)`), as this is supported execution behaviour.

- Fixes a couple of bugs with the Autograd and TensorFlow batch interfaces

- Adds tests

**Benefits:** returning `expval(H)` from QNodes is now supported within the new batch-differentiable pipeline.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
[ch8557]

File: pennylane/gradients/parameter_shift.py [hunk 1]
```diff
@@ -150,19 +139,40 @@ def expval_param_shift(tape, argnum=None, shift=np.pi / 2, gradient_recipes=None
...
+            gradient_coeffs.append(np.array([1.0]))
+            fns.append(h_fn)
             continue
 
         # get the gradient recipe for the trainable parameter
         recipe = gradient_recipes[argnum.index(idx)]
         recipe = recipe or _get_operation_recipe(tape, idx, shift=shift)
         recipe = _process_gradient_recipe(recipe)
         coeffs, multipliers, shifts = recipe
+        fns.append(None)
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift.py:175]: Why is this appending `None`?

File: pennylane/gradients/hamiltonian_grad.py [hunk 2]
```diff
@@ -25,18 +25,35 @@ def hamiltonian_grad(tape, idx, params=None):
...
-    p_idx = tape._par_info[t_idx]["p_idx"]
-
-    new_tape = tape.copy(copy_operations=True, tape_cls=qml.tape.QuantumTape)
     if params is not None:
         new_tape.set_parameters(params=params)
 
-    new_tape._measurements = [qml.expval(op.ops[p_idx])]
+    # get position in queue
+
+    for i, m in enumerate(tape.measurements):
```
Comment (glassnotes) [inline pennylane/gradients/hamiltonian_grad.py:36]: Is there a way to do this using `index` or something, to avoid the `for` loop?

File: tests/gradients/test_parameter_shift.py [hunk 3]
```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
...
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(H)
+
+        a, b, c = coeffs
+        x, y = weights
+        tape.trainable_params = {0, 1, 2, 4}
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift.py:1003]: Any reason it is not using all 3 coeffs of H?

File: tests/gradients/test_parameter_shift.py [hunk 4]
```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
...
+
+        weights = np.array([0.4, 0.5])
+        x, y = weights
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(H1)
+            qml.expval(H2)
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift.py:1048]: This is going to be such a useful feature!

File: tests/gradients/test_parameter_shift.py [hunk 5]
```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
...
+                np.cos(x),
+                np.cos(x) * np.sin(y),
+                -(np.sin(x) * np.sin(y)),
+                0,
+            ],
+            [-d * np.sin(x), 0, 0, 0, 0, np.cos(x)],
+        ]
+
+    def test_autograd(self, tol):
+        """Test gradient of  multiple trainable Hamiltonian coefficients
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift.py:1117]:
```suggestion
        """Test gradient of multiple trainable Hamiltonian coefficients
```

File: tests/gradients/test_parameter_shift.py [hunk 6]
```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
...
+        expected = self.cost_fn_expected(weights, coeffs1, coeffs2)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # second derivative wrt to Hamiltonian coefficients should be zero
+        res = qml.jacobian(self.cost_fn)(weights, coeffs1, coeffs2, dev=dev)
+        assert np.allclose(res[1][:, 2:5], np.zeros([2, 3, 3]), atol=tol, rtol=0)
+        assert np.allclose(res[2][:, -1], np.zeros([2, 1, 1]), atol=tol, rtol=0)
+
+    def test_tf(self, tol):
+        """Test gradient of  multiple trainable Hamiltonian coefficients
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift.py:1134]:
```suggestion
        """Test gradient of multiple trainable Hamiltonian coefficients
```

File: tests/gradients/test_parameter_shift.py [hunk 7]
```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
...
+        expected = self.cost_fn_expected(weights.numpy(), coeffs1.numpy(), coeffs2.numpy())
+        assert np.allclose(jac, expected, atol=tol, rtol=0)
+
+        # second derivative wrt to Hamiltonian coefficients should be zero
+        hess = t.jacobian(jac, [coeffs1, coeffs2])
+        assert np.allclose(hess[0][:, 2:5], np.zeros([2, 3, 3]), atol=tol, rtol=0)
+        assert np.allclose(hess[1][:, -1], np.zeros([2, 1, 1]), atol=tol, rtol=0)
+
+    def test_torch(self, tol):
+        """Test gradient of  multiple trainable Hamiltonian coefficients
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift.py:1156]:
```suggestion
        """Test gradient of multiple trainable Hamiltonian coefficients
```

File: tests/gradients/test_parameter_shift.py [hunk 8]
```diff
@@ -914,3 +914,286 @@ def cost_fn(x):
...
+
+        # second derivative wrt to Hamiltonian coefficients should be zero
+        hess = torch.autograd.functional.jacobian(
+            lambda *args: self.cost_fn(*args, dev=dev), (weights, coeffs1, coeffs2)
+        )
+        assert np.allclose(hess[1][:, 2:5], np.zeros([2, 3, 3]), atol=tol, rtol=0)
+        assert np.allclose(hess[2][:, -1], np.zeros([2, 1, 1]), atol=tol, rtol=0)
+
+    def test_jax(self, tol):
+        """Test gradient of  multiple trainable Hamiltonian coefficients
```
Comment (glassnotes) [inline tests/gradients/test_parameter_shift.py:1180]:
```suggestion
        """Test gradient of multiple trainable Hamiltonian coefficients
```

File: tests/interfaces/test_batch_autograd.py [hunk 9]
```diff
@@ -808,3 +808,93 @@ def cost_fn(x):
...
+    def cost_fn_expected(weights, coeffs1, coeffs2):
+        """Analytic jacobian of cost_fn above"""
+        a, b, c = coeffs1
+        d = coeffs2[0]
+        x, y = weights
+        return [-c * np.sin(x) * np.sin(y) + np.cos(x) * (a + b * np.sin(y)), d * np.cos(x)]
+
+    @staticmethod
+    def cost_fn_jacobian(weights, coeffs1, coeffs2):
+        """Analytic jacobian of cost_fn above"""
```
Comment (glassnotes) [inline tests/interfaces/test_batch_autograd.py:856]: This docstring is identical to the one in the previous function, should one of them be different?

(Same comment for Torch/TF)

**Review by glassnotes — APPROVED** [review]
@josh146 only minor comments, this will be such a nice feature to have! :100:

**Review by anthayes92 — APPROVED** [review]

**Comment by anthayes92** [comment]
Awesome addition @josh146 ! Very thorough testing

File: tests/interfaces/test_batch_torch.py [hunk 10]
```diff
@@ -997,3 +997,101 @@ def cost_fn(x):
...
+                    np.cos(x),
+                    np.cos(x) * np.sin(y),
+                    -(np.sin(x) * np.sin(y)),
+                    0,
+                ],
+                [-d * np.sin(x), 0, 0, 0, 0, np.cos(x)],
+            ]
+        )
+
+    def test_multiple_hamiltonians_not_trainable(self, cost_fn, execute_kwargs, tol):
```
Comment (soranjh) [inline tests/interfaces/test_batch_torch.py:1063]: Is this test added for jax?

File: tests/interfaces/test_batch_torch.py [hunk 11]
```diff
@@ -997,3 +997,101 @@ def cost_fn(x):
...
+
+        res = torch.hstack(
+            torch.autograd.functional.jacobian(
+                lambda *x: cost_fn(*x, dev=dev), (weights, coeffs1, coeffs2)
+            )
+        )
+        expected = self.cost_fn_jacobian(weights, coeffs1, coeffs2)
+        assert np.allclose(res.detach(), expected, atol=tol, rtol=0)
+
+    def test_multiple_hamiltonians_trainable(self, cost_fn, execute_kwargs, tol):
```
Comment (soranjh) [inline tests/interfaces/test_batch_torch.py:1081]: Is this test added for jax?

**Review by soranjh — APPROVED** [review]

File: pennylane/gradients/hamiltonian_grad.py (hunk 2 above)
Comment (josh146) [reply pennylane/gradients/hamiltonian_grad.py:36]: oh good idea 🤔 

Considering that the list of measurements is probably O(1) - O(10), and `list.index` has [linear time complexity](https://stackoverflow.com/questions/5913671/complexity-of-list-indexx-in-python) I would expect both the `for`-`break` and `index` to take about the same amount of time, but indexing seems faster:

image

File: pennylane/gradients/hamiltonian_grad.py [hunk 12]
```diff
@@ -25,18 +25,35 @@ def hamiltonian_grad(tape, idx, params=None):
...
+        if op is m.obs:
+            new_tape._measurements[i] = qml.expval(op.ops[p_idx])
+            queue_position = i
+            break
```
Comment (josh146) [inline pennylane/gradients/hamiltonian_grad.py:40]:
```suggestion
    queue_position = tape.observables.index(op)
    new_tape._measurements[queue_position] = qml.expval(op.ops[p_idx])
```

File: pennylane/gradients/parameter_shift.py (hunk 1 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift.py:175]: Yep, exactly 🙁 

In this branch of the logic, there is no function to append, but then I can't use `zip()` later if the `fn` list is a different length to `shapes` 🙁 

So basically the code is less clean here, to allow for cleaner code later

File: pennylane/tape/tape.py [hunk 13]
```diff
@@ -738,6 +738,29 @@ def trainable_params(self, param_indices):
 
         self._trainable_params = param_indices
 
+    def get_operation(self, idx):
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/tape/tape.py:741]: Yes, was chatting with @mariaschuld last week, and we noticed we were copying these same three lines _everywhere_!

File: tests/gradients/test_parameter_shift.py (hunk 3 above)
Comment (josh146) [reply tests/gradients/test_parameter_shift.py:1003]: Oh... very good question 🤔 I remember doing this deliberately, but for the life of me cannot remember the reason!

Comment (josh146) [reply tests/gradients/test_parameter_shift.py:1003]: Oh! Sorry I misread this line. At first I thought it was `==`, but now I realize it is an assignment `=`. 

Basically, I want to test the case where not all Hamiltonian coefficients are trainable, just to make the test more complicated.

File: tests/interfaces/test_batch_autograd.py (hunk 9 above)
Comment (josh146) [reply tests/interfaces/test_batch_autograd.py:856]: Oops, one above is incorrect 🤦

File: tests/interfaces/test_batch_autograd.py [hunk 14]
```diff
@@ -808,3 +808,93 @@ def cost_fn(x):
...
+                qml.expval(H1)
+                qml.expval(H2)
+
+            return execute([tape], dev, **execute_kwargs)[0]
+
+        return _cost_fn
+
+    @staticmethod
+    def cost_fn_expected(weights, coeffs1, coeffs2):
+        """Analytic jacobian of cost_fn above"""
```
Comment (josh146) [inline tests/interfaces/test_batch_autograd.py:848]:
```suggestion
        """Analytic value of cost_fn above"""
```

File: tests/interfaces/test_batch_tensorflow.py [hunk 15]
```diff
@@ -848,3 +848,101 @@ def test_max_diff(self, tol):
...
+                qml.expval(H1)
+                qml.expval(H2)
+
+            return execute([tape], dev, **execute_kwargs)[0]
+
+        return _cost_fn
+
+    @staticmethod
+    def cost_fn_expected(weights, coeffs1, coeffs2):
+        """Analytic jacobian of cost_fn above"""
```
Comment (josh146) [inline tests/interfaces/test_batch_tensorflow.py:888]:
```suggestion
        """Analytic value of cost_fn above"""
```

File: tests/interfaces/test_batch_torch.py [hunk 16]
```diff
@@ -997,3 +997,101 @@ def cost_fn(x):
...
+                qml.expval(H1)
+                qml.expval(H2)
+
+            return execute([tape], dev, **execute_kwargs)[0]
+
+        return _cost_fn
+
+    @staticmethod
+    def cost_fn_expected(weights, coeffs1, coeffs2):
+        """Analytic jacobian of cost_fn above"""
```
Comment (josh146) [inline tests/interfaces/test_batch_torch.py:1037]:
```suggestion
        """Analytic value of cost_fn above"""
```

File: tests/interfaces/test_batch_torch.py (hunk 10 above)
Comment (josh146) [reply tests/interfaces/test_batch_torch.py:1063]: Not yet, we won't have JAX support until #1573 is merged

File: tests/interfaces/test_batch_torch.py (hunk 11 above)
Comment (josh146) [reply tests/interfaces/test_batch_torch.py:1081]: Not yet, we won't have JAX support until #1573 is merged

File: pennylane/tape/tape.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/tape/tape.py:741]: @anthayes92 that was my original name 😆 

Oddly, I found myself getting confused while writing this, and decided it was better to mirror the existing `tape.get_parameters()` method (which by default only gets trainable parameters).

File: pennylane/gradients/hamiltonian_grad.py (hunk 2 above)
Comment (josh146) [reply pennylane/gradients/hamiltonian_grad.py]: I've made this change :)
